daffodil-udf/src/main/java/org/apache/daffodil/udf/UDFunctionProvider.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.udf;
+
+public abstract class UDFunctionProvider {
+	private Class<?>[] functionClasses;
+
+	public Class<?>[] getFunctionClasses() {
+		return functionClasses;
+	}
+
+	public void setFunctionClasses(Class<?>[] functionClasses) {
+		this.functionClasses = functionClasses;
+	}
+
+	public abstract Object lookupFunctionClass(String namespace, String name);
```

Comment:
Does it make sense to provide a default lookupFunctionClass that iterates over the result of getFunctionClasses, inspects the annotation to find the right class, and calls newInstance if found? For functions that don't require state and have an empty constructor (probably most of them) this would make it easier to create a new UDFProvider. And if a provider did want to provide state to one of their functions, they could override it and manually create their new instances.

Reply:
Hmm, my only concern with this is calling newInstance on a class without an empty constructor causes an exception. Is this okay to do as long as we include a javadoc saying they need to override this function if their class doesn't have an empty constructor?

Reply:
...also just propagating the exception up the chain if it's called on a class with a nonempty constructor?

Reply:
Yeah, throwing an exception/SDE in default lookupFunctionClass saying that the UDF cannot be instantiated seems reasonable to me. Along with the javadoc, this would alert the UDF creator that they need a custom lookup function, or a empty constructor.

A nice benefit about this is a UDFProvider could still take advantage of the function if only one UDF required state, e.g.:
```scala
override def lookupFunctionClass(ns: String, name: String): Object = {
  if (name == "udfThatRequiresState") {
    new UDFThatRequiresState(state)
  } else {
    super.lookupFunctionClass(ns, name)
  }
}
```

Reply:
Roger that, Done.